File: core/ui/shell.py
#!/usr/bin/env python3
"""
AI-OS Shell Interface
GTK-based graphical shell for AI-OS.
"""

import os
import sys
import json
import socket
import struct
import threading
from datetime import datetime

# Try GTK first
try:
    import gi
    gi.require_version('Gtk', '4.0')
    from gi.repository import Gtk, Gdk, GLib, Pango
    HAS_GTK = True
except ImportError:
    HAS_GTK = False


class AgentClient:
    """Client for communicating with AI Agent daemon"""
    
    SOCKET_PATH = "/run/aios/agent.sock"
    
    def send(self, message: dict) -> dict:
        """Send message to agent and get response"""
        try:
            with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
                sock.connect(self.SOCKET_PATH)
                
                # Send message
                data = json.dumps(message).encode('utf-8')
                sock.sendall(struct.pack('!I', len(data)))
                sock.sendall(data)
                
                # Read response
                length_data = sock.recv(4)
                if not length_data:
                    return {'error': 'No response'}
                
                length = struct.unpack('!I', length_data)[0]
                response_data = sock.recv(length)
                
                return json.loads(response_data.decode('utf-8'))
        except Exception as e:
            return {'error': str(e)}
    
    def chat(self, text: str) -> dict:
        return self.send({'cmd': 'chat', 'text': text})
    
    def status(self) -> dict:
        return self.send({'cmd': 'status'})


if HAS_GTK:
    
    class AIosShell(Gtk.Application):
        """Main AI-OS Shell application"""
        
        def __init__(self):
            super().__init__(application_id='com.aios.shell')
            self.agent = AgentClient()
            self.window = None
            
        def do_activate(self):
            if not self.window:
                self.window = ShellWindow(self)
            self.window.present()
    
    
    class ShellWindow(Gtk.ApplicationWindow):
        """Main shell window"""
        
        def __init__(self, app):
            super().__init__(application=app, title="AI-OS")
            self.agent = app.agent
            
            # Make fullscreen
            self.set_decorated(False)
            self.fullscreen()
            
            # Set dark theme
            self.apply_theme()
            
            # Build UI
            self.build_ui()
            
            # Start clock update
            GLib.timeout_add_seconds(1, self.update_clock)
        
        def apply_theme(self):
            """Apply AI-OS theme"""
            css_provider = Gtk.CssProvider()
            css_provider.load_from_data(b"""
                window {
                    background: linear-gradient(180deg, #1a1a2e 0%, #16213e 100%);
                }
                .time-label {
                    font-size: 72px;
                    font-weight: 300;
                    color: white;
                }
                .date-label {
                    font-size: 24px;
                    color: rgba(255, 255, 255, 0.7);
                }
                .agent-input {
                    font-size: 18px;
                    padding: 16px 24px;
                    border-radius: 30px;
                    background: rgba(255, 255, 255, 0.1);
                    border: none;
                    color: white;
                    caret-color: #667eea;
                }
                .agent-input:focus {
                    outline: none;
                    background: rgba(255, 255, 255, 0.15);
                }
                .response-label {
                    font-size: 16px;
                    color: rgba(255, 255, 255, 0.8);
                    padding: 20px;
                }
                .status-bar {
                    background: rgba(0, 0, 0, 0.3);
                    padding: 8px 16px;
                }
                .status-text {
                    font-size: 14px;
                    color: white;
                }
            """)
            
            Gtk.StyleContext.add_provider_for_display(
                Gdk.Display.get_default(),
                css_provider,
                Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION
            )
        
        def build_ui(self):
            """Build the UI"""
            # Main container
            main_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
            self.set_child(main_box)
            
            # Status bar
            status_bar = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL)
            status_bar.add_css_class('status-bar')
            
            self.status_label = Gtk.Label(label="AI-OS")
            self.status_label.add_css_class('status-text')
            status_bar.append(self.status_label)
            
            status_bar_spacer = Gtk.Box()
            status_bar_spacer.set_hexpand(True)
            status_bar.append(status_bar_spacer)
            
            self.clock_mini = Gtk.Label()
            self.clock_mini.add_css_class('status-text')
            status_bar.append(self.clock_mini)
            
            main_box.append(status_bar)
            
            # Center content
            center_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
            center_box.set_valign(Gtk.Align.CENTER)
            center_box.set_halign(Gtk.Align.CENTER)
            center_box.set_vexpand(True)
            center_box.set_spacing(20)
            
            # Time display
            self.time_label = Gtk.Label()
            self.time_label.add_css_class('time-label')
            center_box.append(self.time_label)
            
            # Date display
            self.date_label = Gtk.Label()
            self.date_label.add_css_class('date-label')
            center_box.append(self.date_label)
            
            # Spacer
            spacer = Gtk.Box()
            spacer.set_size_request(-1, 50)
            center_box.append(spacer)
            
            # Agent input
            input_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL)
            input_box.set_halign(Gtk.Align.CENTER)
            input_box.set_size_request(500, -1)
            
            self.input_entry = Gtk.Entry()
            self.input_entry.set_placeholder_text("Ask AI-OS anything...")
            self.input_entry.add_css_class('agent-input')
            self.input_entry.set_hexpand(True)
            self.input_entry.connect('activate', self.on_input_activate)
            input_box.append(self.input_entry)
            
            center_box.append(input_box)
            
            # Response area
            self.response_label = Gtk.Label()
            self.response_label.add_css_class('response-label')
            self.response_label.set_wrap(True)
            self.response_label.set_max_width_chars(60)
            center_box.append(self.response_label)
            
            main_box.append(center_box)
            
            # Update clock immediately
            self.update_clock()
        
        def update_clock(self):
            """Update clock display"""
            now = datetime.now()
            self.time_label.set_text(now.strftime("%H:%M"))
            self.date_label.set_text(now.strftime("%A, %B %d, %Y"))
            self.clock_mini.set_text(now.strftime("%H:%M:%S"))
            return True
        
        def on_input_activate(self, entry):
            """Handle input entry"""
            text = entry.get_text().strip()
            if not text:
                return
            
            entry.set_text("")
            self.response_label.set_text("Processing...")
            
            # Process in background
            def process():
                response = self.agent.chat(text)
                
                if 'error' in response:
                    result = f"Error: {response['error']}"
                else:
                    result = response.get('response', 'No response')
                    
                    # Show action result if any
                    action_result = response.get('action_result')
                    if action_result:
                        if action_result.get('success'):
                            result += f"\n✓ {action_result.get('message', 'Action completed')}"
                        else:
                            result += f"\n✗ {action_result.get('message', 'Action failed')}"
                
                GLib.idle_add(self.response_label.set_text, result)
            
            thread = threading.Thread(target=process)
            thread.daemon = True
            thread.start()


def run_gtk_shell():
    """Run GTK-based shell"""
    app = AIosShell()
    app.run([])


def run_terminal_shell():
    """Run terminal-based shell"""
    print("=" * 50)
    print("     AI-OS Interactive Shell")
    print("=" * 50)
    print("Type 'help' for commands, 'exit' to quit.")
    print()
    
    agent = AgentClient()
    
    while True:
        try:
            user_input = input("AI-OS> ").strip()
            
            if not user_input:
                continue
            
            if user_input.lower() == 'exit':
                print("Goodbye!")
                break
            
            if user_input.lower() == 'help':
                print("""
Available Commands:
  <any text>    - Chat with AI agent
  status        - Show system status  
  clear         - Clear screen
  exit          - Exit shell
  
Examples:
  turn up brightness
  what's the battery level?
  open firefox
  set volume to 50%
""")
                continue
            
            if user_input.lower() == 'clear':
                os.system('clear' if os.name != 'nt' else 'cls')
                continue
            
            if user_input.lower() == 'status':
                response = agent.status()
                if 'error' in response:
                    print(f"Error: {response['error']}")
                else:
                    print(f"Running: {response.get('running')}")
                    print(f"AI Configured: {response.get('ai_configured')}")
                    system = response.get('system', {})
                    print(f"Hostname: {system.get('hostname')}")
                    print(f"Kernel: {system.get('kernel')}")
                    print(f"CPU: {system.get('cpu', 'Unknown')}")
                    print(f"Memory: {system.get('memory_free_mb', 0)}/{system.get('memory_mb', 0)} MB")
                continue
            
            # Send to agent
            response = agent.chat(user_input)
            
            if 'error' in response:
                print(f"Error: {response['error']}")
            else:
                print(f"\n{response.get('response', 'No response')}")
                
                action_result = response.get('action_result')
                if action_result:
                    if action_result.get('success'):
                        print(f"✓ {action_result.get('message', 'Action completed')}")
                    else:
                        print(f"✗ {action_result.get('message', 'Action failed')}")
                    
                    # Show info data if present
                    data = action_result.get('data')
                    if data:
                        for key, value in data.items():
                            print(f"  {key}: {value}")
            
            print()
            
        except KeyboardInterrupt:
            print("\nUse 'exit' to quit.")
        except EOFError:
            break


def main():
    """Main entry point"""
    # Check for display
    if os.environ.get('WAYLAND_DISPLAY') or os.environ.get('DISPLAY'):
        if HAS_GTK:
            run_gtk_shell()
        else:
            print("GTK not available, falling back to terminal")
            run_terminal_shell()
    else:
        run_terminal_shell()


if __name__ == '__main__':
    main()

File: core/ui/shell.c
/**
 * AI-OS Shell UI
 * GTK4-based desktop shell in C.
 * 
 * Compile: gcc -o aios-shell shell.c `pkg-config --cflags --libs gtk4` -lpthread
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <pthread.h>
#include <gtk/gtk.h>

#define AGENT_SOCKET "/run/aios/agent.sock"

/* ==================== Globals ==================== */

static GtkWidget *g_window;
static GtkWidget *g_clock_label;
static GtkWidget *g_date_label;
static GtkWidget *g_input_entry;
static GtkWidget *g_response_label;
static GtkWidget *g_status_bar;

/* ==================== Agent Communication ==================== */

static int send_to_agent(const char *text, char *response, size_t response_size) {
    int sock = socket(AF_UNIX, SOCK_STREAM, 0);
    if (sock < 0) return -1;
    
    struct sockaddr_un addr = {0};
    addr.sun_family = AF_UNIX;
    strcpy(addr.sun_path, AGENT_SOCKET);
    
    if (connect(sock, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        close(sock);
        strcpy(response, "Agent not running");
        return -1;
    }
    
    char msg[2048];
    snprintf(msg, sizeof(msg), "{\"cmd\":\"chat\",\"text\":\"%s\"}", text);
    
    uint32_t len = htonl(strlen(msg));
    send(sock, &len, 4, 0);
    send(sock, msg, strlen(msg), 0);
    
    if (recv(sock, &len, 4, 0) != 4) {
        close(sock);
        return -1;
    }
    len = ntohl(len);
    
    if (len < response_size) {
        recv(sock, response, len, 0);
        response[len] = '\0';
    }
    
    close(sock);
    return 0;
}

/* ==================== Clock Update ==================== */

static gboolean update_clock(gpointer data) {
    time_t now = time(NULL);
    struct tm *tm = localtime(&now);
    
    char time_str[32], date_str[64];
    strftime(time_str, sizeof(time_str), "%H:%M", tm);
    strftime(date_str, sizeof(date_str), "%A, %B %d", tm);
    
    gtk_label_set_text(GTK_LABEL(g_clock_label), time_str);
    gtk_label_set_text(GTK_LABEL(g_date_label), date_str);
    
    return G_SOURCE_CONTINUE;
}

/* ==================== Input Handling ==================== */

static void on_input_activate(GtkEntry *entry, gpointer user_data) {
    GtkEntryBuffer *buffer = gtk_entry_get_buffer(entry);
    const char *text = gtk_entry_buffer_get_text(buffer);
    
    if (!text || !text[0]) return;
    
    gtk_label_set_text(GTK_LABEL(g_response_label), "Thinking...");
    
    char response[4096] = {0};
    if (send_to_agent(text, response, sizeof(response)) == 0) {
        /* Extract response text */
        char *p = strstr(response, "\"response\":\"");
        if (p) {
            p += 12;
            char *end = strchr(p, '"');
            if (end) {
                *end = '\0';
                gtk_label_set_text(GTK_LABEL(g_response_label), p);
            }
        } else {
            gtk_label_set_text(GTK_LABEL(g_response_label), response);
        }
    } else {
        gtk_label_set_text(GTK_LABEL(g_response_label), response);
    }
    
    gtk_entry_buffer_set_text(buffer, "", 0);
}

/* ==================== CSS Styling ==================== */

static void load_css(void) {
    GtkCssProvider *provider = gtk_css_provider_new();
    
    const char *css = 
        "window {"
        "  background: linear-gradient(135deg, #1a1a2e 0%, #16213e 100%);"
        "}"
        ".clock {"
        "  font-size: 96px;"
        "  font-weight: 200;"
        "  color: white;"
        "  text-shadow: 0 2px 10px rgba(0,0,0,0.3);"
        "}"
        ".date {"
        "  font-size: 24px;"
        "  color: #a0a0c0;"
        "}"
        ".ai-input {"
        "  background: rgba(255,255,255,0.1);"
        "  border: 2px solid rgba(255,255,255,0.2);"
        "  border-radius: 25px;"
        "  padding: 12px 24px;"
        "  font-size: 18px;"
        "  color: white;"
        "  min-width: 500px;"
        "}"
        ".ai-input:focus {"
        "  border-color: #667eea;"
        "  box-shadow: 0 0 20px rgba(102,126,234,0.3);"
        "}"
        ".response {"
        "  font-size: 20px;"
        "  color: #c0c0e0;"
        "  padding: 20px;"
        "}"
        ".status-bar {"
        "  background: rgba(0,0,0,0.3);"
        "  padding: 8px 16px;"
        "}"
        ".status-item {"
        "  color: #a0a0c0;"
        "  font-size: 14px;"
        "  margin: 0 8px;"
        "}";
    
    gtk_css_provider_load_from_string(provider, css);
    
    GdkDisplay *display = gdk_display_get_default();
    gtk_style_context_add_provider_for_display(display, 
        GTK_STYLE_PROVIDER(provider), GTK_STYLE_PROVIDER_PRIORITY_APPLICATION);
}

/* ==================== UI Building ==================== */

static void build_status_bar(GtkWidget *parent) {
    GtkWidget *bar = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 0);
    gtk_widget_add_css_class(bar, "status-bar");
    gtk_widget_set_valign(bar, GTK_ALIGN_END);
    
    /* Left section */
    GtkWidget *left = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 8);
    gtk_widget_set_hexpand(left, TRUE);
    gtk_widget_set_halign(left, GTK_ALIGN_START);
    
    GtkWidget *logo = gtk_label_new("🤖 AI-OS");
    gtk_widget_add_css_class(logo, "status-item");
    gtk_box_append(GTK_BOX(left), logo);
    
    gtk_box_append(GTK_BOX(bar), left);
    
    /* Right section */
    GtkWidget *right = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 16);
    gtk_widget_set_halign(right, GTK_ALIGN_END);
    
    GtkWidget *wifi = gtk_label_new("📶 Connected");
    gtk_widget_add_css_class(wifi, "status-item");
    gtk_box_append(GTK_BOX(right), wifi);
    
    GtkWidget *battery = gtk_label_new("🔋 100%");
    gtk_widget_add_css_class(battery, "status-item");
    gtk_box_append(GTK_BOX(right), battery);
    
    GtkWidget *volume = gtk_label_new("🔊 80%");
    gtk_widget_add_css_class(volume, "status-item");
    gtk_box_append(GTK_BOX(right), volume);
    
    gtk_box_append(GTK_BOX(bar), right);
    
    gtk_box_append(GTK_BOX(parent), bar);
    g_status_bar = bar;
}

static void activate(GtkApplication *app, gpointer user_data) {
    load_css();
    
    /* Main window */
    g_window = gtk_application_window_new(app);
    gtk_window_set_title(GTK_WINDOW(g_window), "AI-OS");
    gtk_window_set_default_size(GTK_WINDOW(g_window), 1920, 1080);
    gtk_window_fullscreen(GTK_WINDOW(g_window));
    
    /* Main layout */
    GtkWidget *main_box = gtk_box_new(GTK_ORIENTATION_VERTICAL, 0);
    gtk_window_set_child(GTK_WINDOW(g_window), main_box);
    
    /* Center content */
    GtkWidget *center = gtk_box_new(GTK_ORIENTATION_VERTICAL, 20);
    gtk_widget_set_vexpand(center, TRUE);
    gtk_widget_set_valign(center, GTK_ALIGN_CENTER);
    gtk_widget_set_halign(center, GTK_ALIGN_CENTER);
    gtk_box_append(GTK_BOX(main_box), center);
    
    /* Clock */
    g_clock_label = gtk_label_new("00:00");
    gtk_widget_add_css_class(g_clock_label, "clock");
    gtk_box_append(GTK_BOX(center), g_clock_label);
    
    /* Date */
    g_date_label = gtk_label_new("Loading...");
    gtk_widget_add_css_class(g_date_label, "date");
    gtk_box_append(GTK_BOX(center), g_date_label);
    
    /* Spacer */
    GtkWidget *spacer = gtk_box_new(GTK_ORIENTATION_VERTICAL, 0);
    gtk_widget_set_size_request(spacer, -1, 60);
    gtk_box_append(GTK_BOX(center), spacer);
    
    /* AI Input */
    g_input_entry = gtk_entry_new();
    gtk_entry_set_placeholder_text(GTK_ENTRY(g_input_entry), "Ask AI anything...");
    gtk_widget_add_css_class(g_input_entry, "ai-input");
    g_signal_connect(g_input_entry, "activate", G_CALLBACK(on_input_activate), NULL);
    gtk_box_append(GTK_BOX(center), g_input_entry);
    
    /* Response */
    g_response_label = gtk_label_new("Press Enter to ask the AI");
    gtk_widget_add_css_class(g_response_label, "response");
    gtk_label_set_wrap(GTK_LABEL(g_response_label), TRUE);
    gtk_label_set_max_width_chars(GTK_LABEL(g_response_label), 80);
    gtk_box_append(GTK_BOX(center), g_response_label);
    
    /* Status bar */
    build_status_bar(main_box);
    
    /* Start clock update */
    g_timeout_add_seconds(1, update_clock, NULL);
    update_clock(NULL);
    
    gtk_window_present(GTK_WINDOW(g_window));
}

/* ==================== Main ==================== */

int main(int argc, char *argv[]) {
    GtkApplication *app = gtk_application_new("com.aios.shell", G_APPLICATION_DEFAULT_FLAGS);
    g_signal_connect(app, "activate", G_CALLBACK(activate), NULL);
    
    int status = g_application_run(G_APPLICATION(app), argc, argv);
    g_object_unref(app);
    
    return status;
}
